### PR TITLE
feat(appactivity): Render app activity in activity feed

### PR DESCRIPTION
**Pull request by JustinHoldstock**
- [x] Relies on https://github.com/box/box-ui-elements/pull/887
- [x] Tests

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @JustinHoldstock has signed the CLA. Thanks for the pull request!

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
+/**
+ * @flow
+ * @file Info component used by Keywords component
```
Comment (jeremypress) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: update desc?

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    created_at,
+    created_by,
+    error,
+    id,
+    isPending,
+    rendered_text,
+    permissions,
+    onDelete = noop,
+}: Props): React.Node => {
+    const createdAtTimestamp = new Date(created_at).getTime();
```
Comment (jeremypress) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: If our date objects expect this format would it be better to manipulate at the API layer?

File: src/utils/fields.js [hunk 3]
```diff
@@ -144,6 +148,15 @@ const COMMENTS_FIELDS_TO_FETCH = [
 // Fields that represent users
 const USER_FIELDS = [FIELD_CREATED_BY, FIELD_MODIFIED_BY, FIELD_OWNED_BY, FIELD_ASSIGNED_TO];
 
+// Fields required to fetch ap activity
```
Comment (jeremypress) [inline src/utils/fields.js]: app

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+import InlineDelete from '../inline-delete';
+import { ReadableTime } from '../../../../components/time';
+import Tooltip from '../../../../components/tooltip';
+import { Link } from '../../../../components/link';
+import messages from '../../../common/messages';
+
+import './AppActivity.scss';
+
+const ONE_HOUR_MS = 3600000; // 60 * 60 * 1000
```
Comment (jeremypress) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: Could be useful to share the threshold with `Comment` in case we ever change the underlying value.

File: src/api/AppActivity.js [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+ * @flow
+ * @file Helper for the box App Activity API
+ * @author Box
+ */
+import MarkerBasedAPI from './MarkerBasedAPI';
+import { ERROR_CODE_DELETE_APP_ACTIVITY, HTTP_STATUS_CODE_NOT_FOUND } from '../constants';
+import { APP_ACTIVITY_FIELDS_TO_FETCH } from '../utils/fields';
+
+class AppActivity extends MarkerBasedAPI {
+    /** @property {TasksAPI} - Placeholder permissions object to determine if app activity can be deleted */
```
Comment (ConradJChan) [inline src/api/AppActivity.js]: Instead of `TasksAPI` should this be `BoxItemPermission`?

File: src/api/AppActivity.js [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    permissions: BoxItemPermission = {};
+
+    /**
+     * Map an entry from the AppActivity API to an AppActivityItem.
+     * occurred_at -> created_at
+     *
+     * @param {Object} item - A single entry in the AppActivity API entiries list
```
Comment (ConradJChan) [inline src/api/AppActivity.js]: `entiries` spelling

File: src/api/AppActivity.js [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+     */
+    errorHandler = (error: $AxiosError<any>): void => {
+        if (this.isDestroyed() && typeof this.errorCallback !== 'function') {
+            return;
+        }
+
+        this.permissions = {};
+        const { response } = error;
+
+        // In the case of a 404, the enterprise does not have App Activities enabled.
```
Comment (ConradJChan) [inline src/api/AppActivity.js]: 404 seems a bit odd, shouldn't it be 403 maybe? 404 makes it seem like the API endpoint doesn't exist

File: src/api/AppActivity.js [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    /**
+     * Delete an app activity item
+     *
+     * @param {BoxItem} file - The Box file that App Activity is on
+     * @param {string} appActivityId - An AppActivity item id
+     * @param {Function} successCallback - The success callback
+     * @param {Function} errorCallback - The error callback
+     */
+    deleteAppActivity({
+        file,
```
Comment (ConradJChan) [inline src/api/AppActivity.js]: Can we pass just the fileId instead of the file object?

File: src/api/AppActivity.js [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        appActivityId,
+        successCallback,
+        errorCallback,
+    }: {
+        appActivityId: string,
+        errorCallback: Function,
+        file: BoxItem,
+        successCallback: Function,
+    }): void {
+        this.errorCode = ERROR_CODE_DELETE_APP_ACTIVITY;
```
Comment (ConradJChan) [inline src/api/AppActivity.js]: I've seen in other areas we pass permissions into the delete* call and then invoke `checkApiCallValidity`, is that a pattern we should use here as well instead of storing the permissions in the API class?

File: src/api/Feed.js [hunk 10]
```diff
@@ -167,9 +173,10 @@ class Feed extends Base {
         this.errorCallback = onError;
         const versionsPromise = this.fetchVersions();
         const commentsPromise = this.fetchComments(permissions);
+        const appActivityPromise = this.fetchAppActivity(permissions);
         const tasksPromise = shouldShowNewTasks ? this.fetchTasksNew() : this.fetchTasks();
 
-        Promise.all([versionsPromise, commentsPromise, tasksPromise]).then(feedItems => {
+        Promise.all([versionsPromise, commentsPromise, tasksPromise, appActivityPromise]).then(feedItems => {
```
Comment (ConradJChan) [inline src/api/Feed.js]: Will the app activity items be rendered in a separate PR?

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+    return (
+        <div
+            className="bcs-app-activity-container"
+            data-resin-target="loaded"
+            data-resin-feature={`app_activity_${templateId}_card`}
+        >
+            <div
+                className={classNames('bcs-app-activity', {
+                    'bcs-is-pending': isPending || error,
```
Comment (ConradJChan) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: What does `isPending` mean app activity? Is it as a pending state when it is being deleted? Is the error state the same as pending state? Would it make more semantic sense if we had a class to distinguish an error state?

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss [hunk 12]
```diff
@@ -0,0 +1,64 @@
+@import '../../../common/variables';
+
+.be {
+    .bcs-app-activity-container {
```
Comment (ConradJChan) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss]: Time to SUIT up :)

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js (hunk 2 above)
Comment (JustinHoldstock) [reply src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: Done in API PR. Thanks

File: src/api/AppActivity.js (hunk 7 above)
Comment (JustinHoldstock) [reply src/api/AppActivity.js]: As discussed in our DM, this is a bug in the API that will be fixed in the future

File: src/api/AppActivity.js (hunk 9 above)
Comment (JustinHoldstock) [reply src/api/AppActivity.js]: The reason we leave it out is because it is possible for the user to be missing permissions, but be able to delete the activity. This is true in the case that the user isn't an editor/owner of the file, but they're the one who made the app activity.

File: src/api/Feed.js (hunk 10 above)
Comment (JustinHoldstock) [reply src/api/Feed.js]: This is the Render PR. Note that it's a branch on top of the API layer branch

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js (hunk 11 above)
Comment (JustinHoldstock) [reply src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: The item will be pending when being deleted. I was following what comments do, in regard to errors. I think what we can do here is take up a task to create UI to better distinguish errors for a particular item in the ActivityFeed.

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss (hunk 12 above)
Comment (JustinHoldstock) [reply src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss]: That'll come in a separate PR

File: .github/ISSUE_TEMPLATE/Bug_report.md [hunk 13]
```diff
@@ -11,20 +11,20 @@ Please fill out the following template so we can reproduce and fix your issue as
 
 **Environment:**
 
-- Elements version:
+-   Elements version:
```
Comment (jstoffan) [inline .github/ISSUE_TEMPLATE/Bug_report.md]: This change doesn't seem related. May need to run `yarn install` again to get updated linting rules.

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+                })}
+            >
+                <img className="bcs-app-activity-icon" alt={`${name} Icon`} src={icon_url} />
+                <span className="bcs-app-activity-content">
+                    <div className="bcs-app-activity-headline">
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: It's generally advised to avoid having block elements (div) inside inline elements (span).

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+                        <span className="bcs-app-activity-app-name">{name}</span>
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: Can we add semantic value to the markup? Might be worth looking into `figure` and `figCaption`, for example.

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+                        {(currentUser.id === created_by.id || canDelete) && !isPending && (
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: Should the user id check be folded into the `canDelete` variable?

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+        &.bcs-is-focused {
+            .bcs-inline-delete-container {
+                opacity: 1;
+            }
+        }
+
+        .bcs-app-activity {
+            display: flex;
+            margin-top: 23px;
+            opacity: .9999;
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss]: Is this needed? Can it be `opacity: 1`?

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss [hunk 18]
```diff
@@ -0,0 +1,64 @@
+@import '../../../common/variables';
+
+.be {
+    .bcs-app-activity-container {
+        position: relative;
+        z-index: 1;
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss]: Is a new stacking context needed?

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+        &:hover,
+        &.bcs-is-focused {
+            .bcs-inline-delete-container {
+                opacity: 1;
+            }
+        }
+
+        .bcs-app-activity {
+            display: flex;
+            margin-top: 23px;
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss]: This seems a bit magic-y. Is this value called for by the design?

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+            .bcs-app-activity-icon {
+                flex-shrink: 0;
+                height: 32px;
+                margin-right: 10px;
+                width: 32px;
+            }
+
+            .bcs-app-activity-content {
+                color: $fours;
+                flex: 1;
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss]: Shorthand flex values can cause issues in IE and Safari. Is this `1 1 auto`, `1 0 auto`, `1 1 100%`, etc.?

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+                min-width: 0;
+
+                .bcs-app-activity-link {
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss]: Do these need hover and focus styles? Should they use the `link` class?

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+                    color: $sevens;
+                    cursor: default;
+                    font-size: 12px;
+                    line-height: 20px;
+                }
+            }
+
+            .bcs-app-activity-headline {
+                display: flex;
+                flex-flow: row wrap;
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss]: I think the default value for `flex-direction` is `row`. Should this be `flex-wrap: wrap`?

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss [hunk 23]
```diff
@@ -0,0 +1,64 @@
+@import '../../../common/variables';
+
+.be {
+    .bcs-app-activity-container {
+        position: relative;
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss]: Is this needed?

File: src/elements/common/messages.js [hunk 24]
```diff
@@ -1103,6 +1103,16 @@ const messages: { [string]: MessageDescriptor } = defineMessages({
         description: 'Label for back button',
         defaultMessage: 'Back',
     },
+    inlineDeleteCancel: {
```
[the remark this answers is not in the document]
Comment (JustinHoldstock) [reply src/elements/common/messages.js]: Great idea

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 25]
```diff
@@ -27,6 +27,7 @@ import './ActivitySidebar.scss';
 type ExternalProps = {
     currentUser?: User,
     getUserProfileUrl?: string => Promise<string>,
+    onAppActivityDelete?: Function,
```
[the remark this answers is not in the document]
Comment (JustinHoldstock) [reply src/elements/content-sidebar/ActivitySidebar.js]: Can's say that was a part of the plans. Removing

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 26]
```diff
@@ -339,6 +340,34 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
...
+     *
+     * @param {Object} args - A subset of the app activity
+     * @return void
+     */
+    deleteAppActivity = ({ id }: { id: string }): void => {
+        const {
+            file,
+            api,
+
+            onAppActivityDelete = noop,
```
[the remark this answers is not in the document]
Comment (JustinHoldstock) [reply src/elements/content-sidebar/ActivitySidebar.js]: Removed

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+        z-index: 1;
+
+        &:hover,
+        &.bcs-is-focused {
+            .bcs-inline-delete-container {
+                opacity: 1;
+            }
+        }
+
+        .bcs-app-activity {
```
[the remark this answers is not in the document]
Comment (JustinHoldstock) [reply src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss]: Going to leave all scss changes to the upcoming change to SUIT

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js (hunk 16 above)
Comment (JustinHoldstock) [reply src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: Sure!

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js (hunk 14 above)
Comment (JustinHoldstock) [reply src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: moved to div

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js [hunk 28]
```diff
@@ -0,0 +1,128 @@
...
+    isPending: boolean,
+    onDelete: Function,
+    permissions?: BoxItemPermission,
+    rendered_text: string,
+};
+
+function parseActivity(htmlString: string): Array<HTMLLinkElement> {
+    const doc = parser.parseFromString(htmlString, 'text/html');
+    // $FlowFixMe
+    return Array.from(doc.body.childNodes);
```
[the remark this answers is not in the document]
Comment (JustinHoldstock) [reply src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: `doc` and`doc.body` always exist from the result of DOMParser.parseFromString().

[the remark this answers is not in the document]
Comment (JustinHoldstock) [reply src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: Arg, even though the above is true, Flow doesn't agree

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js (hunk 15 above)
Comment (JustinHoldstock) [reply src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: Might be worth looking at, in the future, but is out of scope of this PR. If you might have noticed, this PR takes a lot of queues from the Comment component, so any changes here are worth making to Comment

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js [hunk 29]
```diff
@@ -0,0 +1,128 @@
...
+    id: string,
+    isPending: boolean,
+    onDelete: Function,
+    permissions?: BoxItemPermission,
+    rendered_text: string,
+};
+
+function parseActivity(htmlString: string): Array<HTMLLinkElement> {
+    const doc = parser.parseFromString(htmlString, 'text/html');
+    // $FlowFixMe
```
[the remark this answers is not in the document]
Comment (JustinHoldstock) [reply src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: Fixed up flow errors.

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss (hunk 21 above)
Comment (JustinHoldstock) [reply src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss]: Already using the `link` class since it's a `Link` component. It requires additional color, as design requested. Note that designs do not outline any different hover behaviour

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss (hunk 17 above)
Comment (JustinHoldstock) [reply src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss]: Trying to be consistent with Comment styling, since this is so similar. Removed it

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss (hunk 18 above)
Comment (JustinHoldstock) [reply src/elements/content-sidebar/activity-feed/app-activity/AppActivity.scss]: Trying to be consistent with Comment styling

File: src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js [hunk 30]
```diff
@@ -0,0 +1,139 @@
...
+import { Link } from '../../../../components/link';
+import messages from '../../../common/messages';
+import { ONE_HOUR_MS } from '../../../../constants';
+
+import './AppActivity.scss';
+
+const parser = new DOMParser();
+
+type Props = {
+    activity_template: Object,
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/app-activity/AppActivity.js]: Do these have types we can leverage?

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 31]
```diff
@@ -339,6 +339,28 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
...
+     * @param {Object} args - A subset of the app activity
+     * @return void
+     */
+    deleteAppActivity = ({ id }: { id: string }): void => {
+        const { file, api } = this.props;
+
+        api.getFeedAPI(false).deleteAppActivity(
+            file,
+            id,
+            () => {
```
Comment (jstoffan) [inline src/elements/content-sidebar/ActivitySidebar.js]: Is the wrapping function still needed?
